data/exporter.go

```diff
@@ -0,0 +1,199 @@
+package data
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/blocknative/dreamboat/beacon"
+	"github.com/blocknative/dreamboat/structs"
+	"github.com/lthibault/log"
+)
+
+var (
+	fileIdleTime   = beacon.DurationPerSlot
+	filesPruneTick = fileIdleTime * 3
+)
+
+type ExportService struct {
+	logger   log.Logger
+	requests chan exportRequest
+}
+
+func NewExportService(logger log.Logger, bufSize int) *ExportService {
+	return &ExportService{logger: logger, requests: make(chan exportRequest, bufSize)}
+}
+
+func (s *ExportService) RunParallel(ctx context.Context, datadir string, numWorkers int) error {
+	for i := 0; i < numWorkers; i++ {
+		go s.Run(ctx, datadir, i)
+	}
+
+	return nil
+}
+
+func (s *ExportService) Run(ctx context.Context, datadir string, id int) {
+	logger := s.logger.WithField("id", id)
+
+	logger.Info("started")
+	defer logger.Info("stopped")
+
+	worker := newWorker(id, datadir, logger)
+	ticker := time.NewTicker(filesPruneTick)
+	for {
+		select {
+		case req := <-s.requests:
+			file, err := worker.getOrCreateFile(req)
+			if err == nil {
+				err = writeToFile(file, req)
+				if err != nil {
+					logger.WithError(err).With(req).Error("failed to write")
+
+					file.Write([]byte("\n"))
+					worker.closeFile(file)
+					logger.WithField("filename", file.Name()).Debug("corrupted file closed")
+				}
+			} else {
+				logger.WithError(err).Error("failed to get/create file: %w", err)
+			}
+
+			select {
+			case req.err <- err: // does not block because it is buffered (1) channel, but better safe than sorry
+			case <-ctx.Done():
+				logger.WithError(ctx.Err()).Error("failed to export request")
+			}
+		case <-ticker.C:
+			worker.closeIdleFiles()
+		case <-ctx.Done():
+			worker.closeIdleFiles()
+			return
+		}
+	}
+}
+
+func (s *ExportService) SubmitGetPayloadRequest(ctx context.Context, gpr structs.SignedBlindedBeaconBlock) error {
+	request := exportRequest{dt: BlockBidAndTraceData, data: gpr.Raw(), slot: gpr.Slot(), timestamp: time.Now(), id: gpr.BlockHash().String(), err: make(chan error, 1)}
+
+	// submit request
+	select {
+	case s.requests <- request:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	// wait for response
+	select {
+	case err := <-request.err:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func writeToFile(file *os.File, req exportRequest) error {
+	// Encode struct as JSON and write to base64 encoder
+	_, err := file.Write([]byte(strconv.Itoa(req.timestamp.Nanosecond())))
```

Comment:
To save some time from this thread - not much, but as this one would be rather hard working - you can create this identifier where you create exportRequest structure.  In this case that would be `relay.go`/`submit.go` -  this way here, you can have it prepared just to be written as one operation :)